doc/design/to_run_stmt.md

```diff
@@ -0,0 +1,178 @@
+WITH param_a = value_a,
+     param_b = value_b
+INTO result_table
+```
+
+The SQLFlow statement above will call the python function `a_python_func` from the file `/run/my_func/data_preprocessor.py` inside the docker image `a_data_scientist/maxcompute_functions:1.0`. The attributes in the `WITH` clause will be passed to the python function as parameters.
+
+## Challenges
+
+- TO RUN can execute functions in Kubernetes and MaxCompute.
+- TO RUN can execute functions in both a single process and a distributed cluster.
+- TO RUN can execute functions built upon various data computing frameworks such as [Dask](https://github.com/dask/dask), [Mars](https://github.com/mars-project/mars), [Ray](https://github.com/ray-project/ray) and so on.
+
+## Design
+
+### Fix the challenges
+
+The semantics of `TO RUN a_data_scientist/maxcompute_functions:1.0/my_func.data_preprocessor.a_python_func` is that `a_python_func` will be executed in a docker container for Kubernetes or a PyODPS node for MaxCompute. The implementation of `a_python_func` is fully customized by users. Because the execution environments for `a_python_func` are different between Kubernetes and MaxCompute, there are some differences considering these three challenges above.
+
+Kubernetes
+
+- Execution environment  
+  `a_python_func` is executed inside the function image `a_data_scientist/maxcompute_functions:1.0`. We can customized this image at the `docker build` stage and make sure that all the dependencies `a_python_func` need are already installed in the image.
+
+- Single process execution  
+  User writes the data transform process using pandas, numpy, tsfresh and so on inside `a_python_func`.
+
+- Distributed cluster execution  
+  User can install Dask/Mars/Ray while building the docker image. Inside the implementation of `a_python_func`, we can create a Dask/Mars/Ray cluster, build a computing graph and submit the distribution execution using the API from these packages.
+
+MaxCompute
+
+- Execution environment  
+  The packages of PyODPS node are already pre-installed and user cannot customize it. `a_python_func` can only use these pre-installed python packages.
+
+- Single process execution
+  Since PyODPS node cannot guarantee it contains all the packages we need such as `tsfresh`, `a_python_func` create a mars cluster containing only one worker and tell this worker to execute another function `another_python_func` to do the data transformation via [Mars remote API](https://github.com/mars-project/mars/issues/1227). The cluster uses our function image `a_data_scientist/maxcompute_functions:1.0`, `another_python_func` and its dependencies are already installed while building this image.
+  
+- Distributed cluster execution  
+  The PyODPS node only installed Mars package, but not Dask and Ray. So we can create a cluster, build the computing graph and submit the distributed execution only using Mars API inside `a_python_func`.
```

Comment:
What is a "computing graph"?

Reply:
Let's take mars for example. After creating mars cluster, user will use mars API to compose the distributed data computing logic. The logic will be translated to a computing graph and submit to the cluster. [link](https://docs.pymars.org/en/latest/distributed/architecture.html)